Box at toplevel in SMT-LIB printer

Halbaroth opened the pull request:
Add a box at the toplevel of the SMT-LIB printer to get a decent output. Replace `hv` box by simple box in the same printer and remove some `Fmt.sp`.

Halbaroth committed:
Box at toplevel in SMT-LIB printer

Add a box at the toplevel of the SMT-LIB printer to get a decent output.
Replace `hv` box by simple box in the same printer and remove some `Fmt.sp`.

bclement-ocp reviewed (approved):
LGTM.

We might want `(Fmt.box pp_silent)` instead of `pp_silent` everywhere as well?

Halbaroth commented:
Uhm it's strange that this PR improves the situation for the SMT-LIB format. Indeed, I didn't put boxes at the top level of the SMT-LIB printer but the AE one. I fixed the issue (and I amend the previous commit because it was it's very purpose, which means you cannot see the wrong commit anymore...)

Maybe the situation is better because I removed the `hv` boxes and some space hints.

Besides, `pp_silent` opens a box at the top level of most of its branches (there are few cases in which we don't want to open a box, for instance for rational numbers). 

Let's imagine we have two printers `pp1` and `pp2`. `pp2` always opens a box at the top level but `pp1` doesn't. Instead, `pp1` always calls `pp2`. We need to open a box in `pp1` too?

bclement-ocp commented:
> Let's imagine we have two printers `pp1` and `pp2`. `pp2` always opens a box at the top level but `pp1` doesn't. Instead, `pp1` always calls `pp2`. We need to open a box in `pp1` too?

No, that should not be necessary!

Halbaroth commented:
So we don't need to open a new box in `pp` or `pp_verbose` because we always do it in `pp_silent` but for some leaves of the AST, right?

Halbaroth committed:
Remove useless boxes

bclement-ocp commented:
> So we don't need to open a new box in `pp` or `pp_verbose` because we always do it in `pp_silent` but for some leaves of the AST, right?

`pp_silent` does not always open a box (e.g. for literals there are no boxes). In any case, opening too many boxes is less of an issue than opening too few.

bclement-ocp commented:
> for literals there are no boxes

Actually, for equalities specifically

Halbaroth commented:
> > for literals there are no boxes
> 
> Actually, for equalities specifically

I didn't make much effort here as we plan to remove the n-ary equality.

Halbaroth committed:
Use a box at the toplevel to be sure

Add a box at the toplevel of `pp_silent` to be sure all expressions
are printed in at least a box.

Halbaroth commented:
I add a `pp_boxed` wrapper around `pp_silent` and use the wrapper version everywhere. `pp_boxed` opens a box to be sure we never print an expression without using a box at the top level ;)

bclement-ocp reviewed (approved):
That also works!
